targets/XPlatCppSdk/source/test/TestApp/TestContext.cpp

```diff
@@ -8,6 +8,29 @@ namespace PlayFabUnit
 {
     void TestContext::EndTest(TestFinishState state, std::string resultMsg)
     {
+        if (finishState != TestFinishState::PENDING)
+        {
+            resultMsg += "You can't mark a test as finished twice.\n";
+            if (finishState == TestFinishState::PASSED)
+            {
+                if (state != TestFinishState::FAILED)
+                {
+                    testResultMsg += resultMsg;
+                    return;
+                }
+                resultMsg = testResultMsg + resultMsg;
+            }
+            else
+            {
+                if (state != TestFinishState::FAILED)
+                {
+                    resultMsg += "You can't mark a test as finished twice.\n";
+                }
+                testResultMsg += resultMsg;
+                return;
+            }
+        }
```

Comment:
It's really hard to follow this logic.

I would prefer:

if (condition && condition)
// logic
else if (condition && condition)
// logic
else if (condition && condition)
// logic
else if (condition
// logic
else
// logic

Basically, your logic above may be right, but if exactly one state is wrong, then we have to reinvent this whole code. The switch statement I describe above may be more code, but it's a lot easier to debug and add a condition pair that may not be covered by your logic.